live-build/misc/migration-scripts/dx_apply

```diff
@@ -16,9 +16,190 @@
 #
 
 #
-# [WIP]
-# The existence of this file is checked by dx_unpack, thus we provide
-# it as an empty file until the needed functionality is added.
+# Creates the dataset layout/hierarchy expected by linux-upgrade for the
+# archive from the migration image to be unpacked. It also takes the
+# required steps and adds an entry to the FreeBSD bootloader so we can
+# optionally boot into Linux.
 #
 
+set -o pipefail
+
+function die() {
+	echo "$(basename "$0"): $*" >&2
+	exit 1
+}
+
+function usage() {
+	echo "$(basename "$0"): $*" >&2
+	echo "Usage: $(basename "$0") <archive-directory>"
+	exit 2
+}
+
+[[ "$(uname -s)" == "SunOS" ]] || die "script can only be used in illumos"
+
+while getopts ':h' opt; do
+	case $opt in
+	h) usage ;;
+	*) usage "Invalid option: -$OPTARG." ;;
+	esac
+done
+
+ARCHIVE_DIR=$1
+[[ -n $ARCHIVE_DIR ]] || usage
+[[ -d $ARCHIVE_DIR ]] || die "$ARCHIVE_DIR is not a directory"
+[[ -f $ARCHIVE_DIR/version.info ]] ||
+	die "$ARCHIVE_DIR does not have a version.info file"
+. "$ARCHIVE_DIR/version.info"
+#
+# WIP: Assert MINIMUM VERSION field for migration.
+#
+
+#
+# Get the RDS and the current ZFS pool that we're currently using.
+#
+RDS=$(mount | awk '/^\/ /{ print $3 }')
+RPOOL=${RDS%%/*}
+
+#
+# Cleanup any previous intermediate state.
+#
+rm -rf /tmp/delphix.*
+rm -f /boot/vmlinuz-* /boot/initrd.img-*
+zfs destroy -r "$RPOOL/ROOT" 2>/dev/null
```

Comment:
We should probably also double check that "$RPOOL/ROOT" doesn't exist after this operation, since we don't check for the failure of `zfs destroy`.